File: src/streaming.zig
//! Streaming JSON parser for Claude's stream-json output format.
//!
//! Parses Claude's streaming JSON events and extracts text content
//! for real-time display.

const std = @import("std");

/// Event types from Claude's stream-json format
pub const EventType = enum {
    system,
    assistant,
    user,
    stream_event,
    result,
    unknown,
};

/// Parsed streaming event
pub const StreamEvent = struct {
    event_type: EventType,
    /// For text_delta events, contains the text chunk
    text: ?[]const u8 = null,
    /// For tool_use events, contains the tool name
    tool_name: ?[]const u8 = null,
    /// For result events, contains the final result
    result: ?[]const u8 = null,
    /// Whether this is an error
    is_error: bool = false,
};

/// Parse a single JSON line from Claude's streaming output
pub fn parseStreamLine(allocator: std.mem.Allocator, line: []const u8) !StreamEvent {
    const parsed = std.json.parseFromSlice(std.json.Value, allocator, line, .{}) catch {
        return .{ .event_type = .unknown };
    };
    defer parsed.deinit();

    const root = parsed.value;
    if (root != .object) return .{ .event_type = .unknown };

    const obj = root.object;

    // Get event type
    const type_str = if (obj.get("type")) |t| switch (t) {
        .string => |s| s,
        else => return .{ .event_type = .unknown },
    } else return .{ .event_type = .unknown };

    const event_type: EventType = if (std.mem.eql(u8, type_str, "stream_event"))
        .stream_event
    else if (std.mem.eql(u8, type_str, "assistant"))
        .assistant
    else if (std.mem.eql(u8, type_str, "user"))
        .user
    else if (std.mem.eql(u8, type_str, "system"))
        .system
    else if (std.mem.eql(u8, type_str, "result"))
        .result
    else
        .unknown;

    var event = StreamEvent{ .event_type = event_type };

    // Handle stream_event with content_block_delta (text streaming)
    if (event_type == .stream_event) {
        if (obj.get("event")) |evt| {
            if (evt == .object) {
                const evt_obj = evt.object;
                if (evt_obj.get("type")) |evt_type| {
                    if (evt_type == .string and std.mem.eql(u8, evt_type.string, "content_block_delta")) {
                        if (evt_obj.get("delta")) |delta| {
                            if (delta == .object) {
                                if (delta.object.get("text")) |text| {
                                    if (text == .string) {
                                        event.text = try allocator.dupe(u8, text.string);
                                    }
                                }
                            }
                        }
                    }
                }
            }
        }
    }

    // Handle assistant message with tool_use
    if (event_type == .assistant) {
        if (obj.get("message")) |msg| {
            if (msg == .object) {
                if (msg.object.get("content")) |content| {
                    if (content == .array and content.array.items.len > 0) {
                        const first = content.array.items[0];
                        if (first == .object) {
                            if (first.object.get("type")) |t| {
                                if (t == .string and std.mem.eql(u8, t.string, "tool_use")) {
                                    if (first.object.get("name")) |name| {
                                        if (name == .string) {
                                            event.tool_name = try allocator.dupe(u8, name.string);
                                        }
                                    }
                                }
                            }
                        }
                    }
                }
            }
        }
    }

    // Handle result
    if (event_type == .result) {
        if (obj.get("result")) |res| {
            if (res == .string) {
                event.result = try allocator.dupe(u8, res.string);
            }
        }
        if (obj.get("is_error")) |is_err| {
            if (is_err == .bool) {
                event.is_error = is_err.bool;
            }
        }
    }

    return event;
}

/// Stream handler callback type
pub const StreamCallback = *const fn (event: StreamEvent) void;

/// Print text delta to stdout (for streaming display)
pub fn printTextDelta(event: StreamEvent) void {
    if (event.text) |text| {
        _ = std.fs.File.stdout().write(text) catch {};
    }
    if (event.tool_name) |name| {
        std.debug.print("\n\x1b[0;36m[TOOL]\x1b[0m {s}\n", .{name});
    }
}

test "parse text delta" {
    const json =
        \\{"type":"stream_event","event":{"type":"content_block_delta","index":0,"delta":{"type":"text_delta","text":"Hello"}}}
    ;

    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();

    const event = try parseStreamLine(arena.allocator(), json);
    try std.testing.expectEqual(EventType.stream_event, event.event_type);
    try std.testing.expectEqualStrings("Hello", event.text.?);
}

test "parse tool use" {
    const json =
        \\{"type":"assistant","message":{"content":[{"type":"tool_use","name":"Bash","id":"123"}]}}
    ;

    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();

    const event = try parseStreamLine(arena.allocator(), json);
    try std.testing.expectEqual(EventType.assistant, event.event_type);
    try std.testing.expectEqualStrings("Bash", event.tool_name.?);
}

File: lib/noface/util/streaming.ex
defmodule Noface.Util.Streaming do
  @moduledoc """
  Streaming JSON parser for Claude's stream-json output format.

  Uses Jaxon for proper streaming JSON parsing that handles incomplete
  chunks and provides event-based parsing for large responses.
  """

  @type event_type :: :system | :assistant | :user | :stream_event | :result | :unknown

  @type stream_event :: %{
          event_type: event_type(),
          text: String.t() | nil,
          tool_name: String.t() | nil,
          tool_input_summary: String.t() | nil,
          result: String.t() | nil,
          is_error: boolean()
        }

  defmodule StreamParser do
    @moduledoc """
    Stateful streaming parser using Jaxon.

    Handles incomplete JSON chunks and emits events as they become complete.
    """

    defstruct buffer: "", events: []

    @type t :: %__MODULE__{
            buffer: String.t(),
            events: [Noface.Util.Streaming.stream_event()]
          }

    @doc "Create a new parser state"
    def new, do: %__MODULE__{}

    @doc "Feed data into the parser, returns {events, new_state}"
    @spec feed(t(), String.t()) :: {[Noface.Util.Streaming.stream_event()], t()}
    def feed(%__MODULE__{buffer: buffer} = state, data) do
      full_buffer = buffer <> data

      # Split on newlines and try to parse each complete line
      {events, remaining} = parse_lines(full_buffer)

      {events, %{state | buffer: remaining, events: []}}
    end

    defp parse_lines(data) do
      lines = String.split(data, "\n")

      # The last element might be incomplete
      {complete_lines, incomplete} =
        case lines do
          [] -> {[], ""}
          [single] -> {[], single}
          many -> {Enum.drop(many, -1), List.last(many)}
        end

      events =
        complete_lines
        |> Enum.reject(&(&1 == ""))
        |> Enum.map(&Noface.Util.Streaming.parse_stream_line/1)

      {events, incomplete}
    end
  end

  @doc """
  Create a new streaming parser.
  """
  @spec new_parser() :: StreamParser.t()
  def new_parser, do: StreamParser.new()

  @doc """
  Feed data into the parser, returns parsed events and new parser state.
  """
  @spec feed_parser(StreamParser.t(), String.t()) :: {[stream_event()], StreamParser.t()}
  def feed_parser(parser, data), do: StreamParser.feed(parser, data)

  @doc """
  Parse a single JSON line from Claude's streaming output.
  """
  @spec parse_stream_line(String.t()) :: stream_event()
  def parse_stream_line(line) do
    # Try Jaxon first for potentially large JSON
    case parse_with_jaxon(line) do
      {:ok, json} -> parse_json_event(json)
      :error -> parse_with_jason(line)
    end
  end

  defp parse_with_jaxon(line) do
    try do
      # Jaxon.decode! for complete JSON objects
      json = Jaxon.decode!(line)
      {:ok, json}
    rescue
      _ -> :error
    end
  end

  defp parse_with_jason(line) do
    case Jason.decode(line) do
      {:ok, json} ->
        parse_json_event(json)

      {:error, _} ->
        %{
          event_type: :unknown,
          text: nil,
          tool_name: nil,
          tool_input_summary: nil,
          result: nil,
          is_error: false
        }
    end
  end

  @doc """
  Parse a stream of data using Jaxon's streaming capabilities.

  This is useful for parsing very large JSON responses incrementally.
  """
  @spec stream_parse(Enumerable.t()) :: Enumerable.t()
  def stream_parse(data_stream) do
    data_stream
    |> Stream.flat_map(fn chunk ->
      chunk
      |> String.split("\n", trim: true)
      |> Enum.map(&parse_stream_line/1)
    end)
  end

  defp parse_json_event(%{"type" => type_str} = json) do
    event_type = parse_event_type(type_str)

    base_event = %{
      event_type: event_type,
      text: nil,
      tool_name: nil,
      tool_input_summary: nil,
      result: nil,
      is_error: false
    }

    base_event
    |> maybe_parse_stream_event(event_type, json)
    |> maybe_parse_assistant(event_type, json)
    |> maybe_parse_result(event_type, json)
  end

  defp parse_json_event(_) do
    %{
      event_type: :unknown,
      text: nil,
      tool_name: nil,
      tool_input_summary: nil,
      result: nil,
      is_error: false
    }
  end

  defp parse_event_type("stream_event"), do: :stream_event
  defp parse_event_type("assistant"), do: :assistant
  defp parse_event_type("user"), do: :user
  defp parse_event_type("system"), do: :system
  defp parse_event_type("result"), do: :result
  defp parse_event_type(_), do: :unknown

  defp maybe_parse_stream_event(event, :stream_event, json) do
    text =
      get_in(json, ["event", "delta", "text"]) ||
        get_in(json, ["event", "delta", "partial_json"])

    %{event | text: text}
  end

  defp maybe_parse_stream_event(event, _, _), do: event

  defp maybe_parse_assistant(event, :assistant, json) do
    case get_in(json, ["message", "content"]) do
      [%{"type" => "tool_use", "name" => name} = tool_content | _] ->
        summary = extract_tool_summary(name, Map.get(tool_content, "input", %{}))
        %{event | tool_name: name, tool_input_summary: summary}

      _ ->
        event
    end
  end

  defp maybe_parse_assistant(event, _, _), do: event

  defp maybe_parse_result(event, :result, json) do
    %{
      event
      | result: json["result"],
        is_error: Map.get(json, "is_error", false)
    }
  end

  defp maybe_parse_result(event, _, _), do: event

  @doc """
  Extract a human-readable summary from tool input based on tool type.
  """
  @spec extract_tool_summary(String.t(), map()) :: String.t() | nil
  def extract_tool_summary(tool_name, input) when tool_name in ["Read", "Edit", "Write"] do
    Map.get(input, "file_path")
  end

  def extract_tool_summary("Bash", input) do
    case Map.get(input, "command") do
      nil ->
        nil

      cmd when byte_size(cmd) <= 60 ->
        cmd

      cmd ->
        String.slice(cmd, 0, 60) <> "..."
    end
  end

  def extract_tool_summary("Glob", input), do: Map.get(input, "pattern")
  def extract_tool_summary("Grep", input), do: Map.get(input, "pattern")
  def extract_tool_summary("Task", input), do: Map.get(input, "description")
  def extract_tool_summary(_, _), do: nil

  @doc """
  Print text delta to stdout (for streaming display).
  """
  @spec print_text_delta(stream_event()) :: :ok
  def print_text_delta(%{text: text}) when is_binary(text) do
    IO.write(text)
    :ok
  end

  def print_text_delta(%{tool_name: name, tool_input_summary: summary})
      when is_binary(name) do
    if summary do
      IO.puts("\n\e[0;36m[TOOL]\e[0m #{name}: #{summary}")
    else
      IO.puts("\n\e[0;36m[TOOL]\e[0m #{name}")
    end

    :ok
  end

  def print_text_delta(_), do: :ok

  @doc """
  Collect all text from a stream of events.
  """
  @spec collect_text([stream_event()]) :: String.t()
  def collect_text(events) do
    events
    |> Enum.map(& &1.text)
    |> Enum.reject(&is_nil/1)
    |> Enum.join("")
  end

  @doc """
  Collect all tool uses from a stream of events.
  """
  @spec collect_tools([stream_event()]) :: [{String.t(), String.t() | nil}]
  def collect_tools(events) do
    events
    |> Enum.filter(& &1.tool_name)
    |> Enum.map(&{&1.tool_name, &1.tool_input_summary})
  end
end
